resource-dependencies (#222)
aws-samples/cfn101-workshop: content/intermediate/templates/resource-dependencies/index.md

Use the AWS CloudFormation Console to [create a stack](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-console-create-stack.html) using the `resource-dependencies-without-dependson.yaml` template:

:::::tabs{variant="container"}

::::tab{id="cloud9" label="Cloud9"}
1. In the **Cloud9 terminal** navigate to `code/workspace/resource-dependencies`:
:::code{language=shell showLineNumbers=false showCopyAction=true}
cd cfn101-workshop/code/workspace/resource-dependencies
:::
1. Use the AWS CLI to create the stack. The required parameter `--template-body` have been pre-filled for you.
:::code{language=shell showLineNumbers=false showCopyAction=true}

> aws cloudformation create-stack --stack-name cfn-workshop-resource-dependencies --template-body file://resource-dependencies-without-dependson.yaml

Reviewer: ```suggestion
aws cloudformation create-stack --stack-name cfn-workshop-resource-dependencies \
--template-body file://resource-dependencies-without-dependson.yaml
```

"StackId": "arn:aws:cloudformation:us-east-1:123456789012:stack/cfn-workshop-resource-dependencies/739fafa0-e4d7-11ed-a000-12d9009553ff"
:::
 1. Open the **[AWS CloudFormation](https://console.aws.amazon.com/cloudformation)** console in a new tab and check if the stack status is **CREATE_COMPLETE**.
::::

::::tab{id="local" label="Local development"}
1. Navigate to the [AWS CloudFormation Console](https://console.aws.amazon.com/cloudformation/).
2. From **Create stack**, choose **With new resources (standard)**.
3. Choose the **Template is ready** option. From **Specify template**, choose **Upload a template file**. Upload the `resource-dependencies-without-dependson.yaml` template, and choose **Next**.
4. Enter a stack name. For example, specify `cfn-workshop-resource-dependencies`. When ready, choose **Next**.
5. Choose to accept default values on the **Configure stack options** page; scroll to the bottom of the page, and choose **Next**.
6. In the **Review** page, scroll to the bottom and choose **Submit**.
::::
:::::

Refresh the page until you see the `CREATE_COMPLETE` status for your stack. Now, let’s review stack events, that should look similar to the image shown next:

> 

Reviewer: Note the following image [resource-dependencies-lab.png] on line 96 has the incorrect stack name in the logical ID column

```

Follow the same steps as above to create a new stack using the `resource-dependencies-with-dependson.yaml` template file. Make sure to provide a different stack name, for example `cfn-workshop-resource-dependencies-dependson`, and create the stack.

This time, your stack events should look different:

> 

Reviewer: Following image [resource-dependencies-lab-dependson.png] has incorrect stack name

Let’s create a stack, and verify the above behavior. Use the AWS CloudFormation Console to [create a stack](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-console-create-stack.html) using the `resource-dependency-with-intrinsic-functions.yaml` template:

:::::tabs{variant="container"}

::::tab{id="cloud9" label="Cloud9"}
1. In the **Cloud9 terminal** navigate to `code/workspace/resource-dependencies`:
:::code{language=shell showLineNumbers=false showCopyAction=true}
cd cfn101-workshop/code/workspace/resource-dependencies
:::
1. Use the AWS CLI to create the stack. The required parameter `--template-body` have been pre-filled for you.
:::code{language=shell showLineNumbers=false showCopyAction=true}

> aws cloudformation create-stack --stack-name cfn-workshop-resource-dependencies-ref-getatt --template-body file://resource-dependencies-with-intrinsic-functions.yaml

Reviewer: ```suggestion
aws cloudformation create-stack --stack-name cfn-workshop-resource-dependencies-ref-getatt \
--template-body file://resource-dependencies-with-intrinsic-functions.yaml
```

:::

> 1. If the `create-stack` command was successfully sent, CloudFormation will return `StackId`.

Reviewer: I got an error `An error occurred (ValidationError) when calling the CreateStack operation: Parameters: [EmailAddress] must have values`
Suggest passing in parameter to the create-stack command `--parameters ParameterKey="EmailAddress",ParameterValue="your-email-address-here`

Let’s create a stack, and verify the above behavior. Use the AWS CloudFormation Console to [create a stack](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-console-create-stack.html) using the `resource-dependency-with-intrinsic-functions.yaml` template:

:::::tabs{variant="container"}

::::tab{id="cloud9" label="Cloud9"}
1. In the **Cloud9 terminal** navigate to `code/workspace/resource-dependencies`:
:::code{language=shell showLineNumbers=false showCopyAction=true}
cd cfn101-workshop/code/workspace/resource-dependencies
:::
1. Use the AWS CLI to create the stack. The required parameter `--template-body` have been pre-filled for you.
:::code{language=shell showLineNumbers=false showCopyAction=true}
aws cloudformation create-stack --stack-name cfn-workshop-resource-dependencies-ref-getatt \

> --template-body file://resource-dependencies-with-intrinsic-functions.yaml --parameters \

Reviewer: ```suggestion
--template-body file://resource-dependencies-with-intrinsic-functions.yaml \
```

> ParameterKey="EmailAddress",ParameterValue="your-email-address-here"

Reviewer: ```suggestion
--parameters ParameterKey="EmailAddress",ParameterValue="your-email-address-here"
```

      - Key: Name
        Value: Resource-dependencies-workshop
```

:::::tabs{variant="container"}
::::tab{id="cloud9" label="Cloud9"}
1. In the **Cloud9 terminal** navigate to `code/workspace/resource-dependencies`:
  :::code{language=shell showLineNumbers=false showCopyAction=true}
  cd cfn101-workshop/code/workspace/resource-dependencies
  :::
1. Use the AWS CLI to create the stack. The required parameter `--template-body` have been pre-filled for you.
  :::code{language=shell showLineNumbers=false showCopyAction=true}

>   aws cloudformation create-stack --stack-name cfn-workshop-resource-dependencies-challenge --template-body file://resource-dependencies-with-intrinsic-functions.yaml

Reviewer: Error due to wrong yaml file being specified: `An error occurred (ValidationError) when calling the CreateStack operation: Parameters: [EmailAddress] must have values`
Also added line break
```suggestion
  aws cloudformation create-stack --stack-name cfn-workshop-resource-dependencies-challenge \
  --template-body file://resource-dependencies-challenge.yaml
```

:::
1. If the `create-stack` command was successfully sent, CloudFormation will return `StackId`.
  :::code{language=shell showLineNumbers=false showCopyAction=false}
  "StackId": "arn:aws:cloudformation:us-east-1:123456789012:stack/cfn-workshop-resource-dependencies-challenge/739fafa0-e4d7-11ed-a000-12d9009553ff"
:::
1. Open the **[AWS CloudFormation](https://console.aws.amazon.com/cloudformation)** console in a new tab and check if the stack status is **CREATE_COMPLETE**.
::::
::::tab{id="local" label="Local development"}
1. Navigate to the [AWS CloudFormation Console](https://console.aws.amazon.com/cloudformation/).
1. From **Create stack**, choose **With new resources (standard)**.

> 1. Choose the **Template is ready** option. From **Specify template**, choose **Upload a template file**. Upload the `resource-dependencies-with-intrinsic-functions.yaml` template, and choose **Next**.

Reviewer: Fix yaml file name
```suggestion
1. Choose the **Template is ready** option. From **Specify template**, choose **Upload a template file**. Upload the `resource-dependencies-challenge.yaml` template, and choose **Next**.
```

> 1. Enter a stack name. For example, `cfn-workshop-resource-dependencies-challenge`.

Reviewer: ```suggestion
1. Enter a stack name. For example, `cfn-workshop-resource-dependencies-challenge` and choose **Next**.
```

> 1. In the **Parameters** section, provide an email address for Amazon SNS topic subscription; when ready, choose **Next**.

Reviewer: ```suggestion
```
